core/src/main/java/com/scalar/db/transaction/consensuscommit/CrudHandler.java

```diff
@@ -196,8 +197,26 @@ Optional<TransactionResult> read(@Nullable Snapshot.Key key, Get get) throws Cru
 
   private Optional<TransactionResult> executeRecovery(
       Snapshot.Key key, Selection selection, TransactionResult result) throws CrudException {
+    RecoveryExecutor.RecoveryType recoveryType;
+    if (snapshot.getIsolation() == Isolation.READ_COMMITTED) {
+      // In READ_COMMITTED isolation
+
+      if (readOnly) {
+        // In read-only mode, we don't recover the record, but return the committed result
+        recoveryType = RecoveryExecutor.RecoveryType.RETURN_COMMITTED_RESULT_AND_NOT_RECOVER;
+      } else {
+        // In read-write mode, we recover the record and return the committed result
+        recoveryType = RecoveryExecutor.RecoveryType.RETURN_COMMITTED_RESULT_AND_RECOVER;
+      }
+    } else {
+      // In SNAPSHOT or SERIALIZABLE isolation, we always recover the record and return the latest
+      // result
+      recoveryType = RecoveryExecutor.RecoveryType.RETURN_LATEST_RESULT_AND_RECOVER;
+    }
```

Comment:
The `RecoveryType` is chosen based in the isolation level and whether the transaction is read-only or not.